compiler/rustc_const_eval/src/transform/check_consts/qualifs.rs

```diff
@@ -155,12 +154,27 @@ impl Qualif for NeedsNonConstDrop {
             return false;
         }
 
-        // FIXME(effects) constness
+        // FIXME(effects): If `destruct` is not a `const_trait`,
+        // or effects are disabled in this crate, then give up.
+        let destruct_def_id = cx.tcx.require_lang_item(LangItem::Destruct, Some(cx.body.span));
+        if cx.tcx.generics_of(destruct_def_id).host_effect_index.is_none()
+            || !cx.tcx.features().effects
+        {
+            return NeedsDrop::in_any_value_of_ty(cx, ty);
+        }
+
         let obligation = Obligation::new(
             cx.tcx,
             ObligationCause::dummy_with_span(cx.body.span),
             cx.param_env,
-            ty::TraitRef::from_lang_item(cx.tcx, LangItem::Destruct, cx.body.span, [ty]),
+            ty::TraitRef::new(
+                cx.tcx,
+                destruct_def_id,
+                [
+                    ty::GenericArg::from(ty),
+                    ty::GenericArg::from(cx.tcx.expected_const_effect_param_for_body(cx.def_id())),
```

Comment:
I just took a look at the docs of this function, it looks like it is trying to say that "inference fails if we don't equate with true" which is incorrect after inference fallback was implemented. I guess the reason now is probably to double-check with different typeck paths and see if they match? Not entirely sure about that. perhaps simply replace the comment suggesting that it is necessary with a FIXME to investigate whether the `host.eq(true)` path should just be eliminated and give us some performance wins.

Reply:
Are you talking about the comments for `expected_const_effect_param_for_body`?

In this case, we should always be creating a `Const::Param` with `sym::host` -- I'm just using this function since it's more convenient.